x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java

```diff
@@ -0,0 +1,307 @@
+                wrappedDataEncryptionKey = wrapKey(dataEncryptionKey, this.masterSecretKey);
+            } catch (NoSuchAlgorithmException | NoSuchPaddingException | InvalidKeyException | IllegalBlockSizeException e) {
+                throw new IOException(e);
+            }
+            try (InputStream stream = new ByteArrayInputStream(wrappedDataEncryptionKey)) {
+                this.encryptionMetadataBlobContainer.writeBlob(blobName, stream, wrappedDataEncryptionKey.length, failIfAlreadyExists);
+            }
+            final long encryptedSize = GCMPacketsCipherInputStream.getEncryptionSizeFromPlainSize(blobSize);
+            try (InputStream encryptedInputStream = GCMPacketsCipherInputStream.getEncryptor(inputStream, dataEncryptionKey, NONCE,
+                    BC_FIPS_PROVIDER)) {
+                this.delegatedBlobContainer.writeBlob(blobName, encryptedInputStream, encryptedSize, failIfAlreadyExists);
+            }
+        }
+
+        @Override
+        public void writeBlobAtomic(String blobName, InputStream inputStream, long blobSize, boolean failIfAlreadyExists)
+                throws IOException {
+            // does not support atomic write
+            writeBlob(blobName, inputStream, blobSize, failIfAlreadyExists);
+        }
+
+        @Override
+        public void deleteBlob(String blobName) throws IOException {
+            this.delegatedBlobContainer.deleteBlob(blobName);
+            this.encryptionMetadataBlobContainer.deleteBlob(blobName);
+        }
+
+        @Override
+        public DeleteResult delete() throws IOException {
+            DeleteResult result = this.delegatedBlobContainer.delete();
+            this.encryptionMetadataBlobContainer.delete();
+            return result;
+        }
+
+        @Override
+        public Map<String, BlobMetaData> listBlobs() throws IOException {
+            return this.delegatedBlobContainer.listBlobs();
+        }
+
+        @Override
+        public Map<String, BlobContainer> children() throws IOException {
+            return this.delegatedBlobContainer.children();
+        }
+
+        @Override
+        public Map<String, BlobMetaData> listBlobsByPrefix(String blobNamePrefix) throws IOException {
+            Map<String, BlobMetaData> delegatedBlobs = this.delegatedBlobContainer.listBlobsByPrefix(blobNamePrefix);
+            Map<String, BlobMetaData> delegatedBlobsWithPlainSize = new HashMap<>(delegatedBlobs.size());
+            for (Map.Entry<String, BlobMetaData> entry : delegatedBlobs.entrySet()) {
+                delegatedBlobsWithPlainSize.put(entry.getKey(), new BlobMetaData() {
+
+                    @Override
+                    public String name() {
+                        return entry.getValue().name();
+                    }
+
+                    @Override
+                    public long length() {
+                        return GCMPacketsCipherInputStream.getDecryptionSizeFromCipherSize(entry.getValue().length());
```

Comment:
Will this actually return the exact unencrypted size? (it doesn't look like it does from the implementation, though it isn't so important that it does anymore since we changed the way we handle file sizes in the incremental snapshotting logic away from listing and towards using the metadata we store ... I'd still like to understand if this breaks being able to rely on these numbers going forward :))

Reply:
Thanks for the feedback Armin!
Yes it does, this implementation can compute the plain size from the cipher size, and vice versa, without reading any information in the metadata blob, because the GCM packets have a fixed size.

I have another alternative implementation inbound that must read the metadata blob to determine the decryption size. Arguably this implementation must also read the metadata blob if we ever want to change the packet size.

I will try not to break the requirement that the size returns the correct size, it's not a difficult requirement.

Reply:
> I will try not to break the requirement that the size returns the correct size, it's not a difficult requirement.

Ok, don't worry about it too much though. We can adjust the Javadoc to account for this as well and be happy if it's a hassle. We don't really need this to be exact any more.

Reply:
To me it feels "more correct" to return the size on disk for the blob, which in this case is the encrypted size. But we also proud ourselves that the encryption is completely transparent, which is not strictly the case if an encrypted repository returns different sizes than a plain one.

I get that you're also on my "more correct" side, so I'll make the change suggested (ie. don't wrap it).